core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -1087,7 +1124,6 @@ public void prepareForIndexRecovery() {
             default:
                 throw new IllegalStateException("No operation defined for [" + operation + "]");
         }
-        ExceptionsHelper.reThrowIfNotNull(result.getFailure());
```

Comment:
TransportResyncAction (which is a TransportWriteAction) has its own way of dealing with these failures. I wanted the failure handling logic to be the same as for TransportShardBulkAction.